lib/bap_llvm/bap_llvm_ogre_loader.ml

```diff
@@ -0,0 +1,72 @@
+open Core_kernel.Std
+open Bap.Std
+open Monads.Std
+open Or_error
+open Bap_llvm_ogre_types
+
+module Make(M : Monad.S) = struct
+  include Ogre.Make(M)
+  type 'a m = 'a M.t
+end
+
+module Dispatch(M : Monad.S) = struct
+  module Fact = Make(Monad.Ident)
+  module Elf = Bap_llvm_ogre_elf.Make(Fact)
+  module Coff = Bap_llvm_ogre_coff.Make(Fact)
+  module Macho = Bap_llvm_ogre_macho.Make(Fact)
+  open Fact.Syntax
+
+  type typ = Elf | Coff | Macho | Unknown [@@deriving sexp]
+
+  let filetype_of_string s =
+    try
+      typ_of_sexp (Sexp.of_string s)
+    with _ -> Unknown
+
+  let image =
+    Ogre.require file_type >>= fun s ->
+    match filetype_of_string s with
+    | Elf -> Elf.image
+    | Coff -> Coff.image
+    | Macho -> Macho.image
+    | Unknown -> Fact.failf "file type is not supported" ()
+
+end
+
+module Loader = struct
+
+  include Dispatch(Monad.Ident)
+
+  exception Llvm_loader_fail of int
+
+  let _ = Callback.register_exception
+      "Llvm_loader_fail" (Llvm_loader_fail 0)
+
+  let to_image_doc doc =
+    match Fact.exec image doc with
+    | Ok doc -> Ok (Some doc)
+    | Error er -> Error er
+
+  let from_data data =
+    try
+      let doc = Bap_llvm_binary.bap_llvm_load data in
+      Ogre.Doc.from_string doc >>= fun doc -> to_image_doc doc
+    with Llvm_loader_fail n -> match n with
+      | 1 -> Ok None
+      | 2 -> Or_error.error_string "file corrupted"
+      | n -> Or_error.errorf "fail with unexpeced error code %d" n
+
+  let from_file path =
```

Comment:
We should think about how to change the loader interface, so that if a `from_file` is not provided, then the image loader will use the `from_data`.